gsk4/src/render_node.rs

```diff
@@ -0,0 +1,161 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use crate::RenderNodeType;
+use glib::translate::*;
+use glib::StaticType;
+use std::fmt;
+use std::ptr;
+
+glib::wrapper! {
+    #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+    pub struct RenderNode(Shared<ffi::GskRenderNode>);
+
+    match fn {
+        ref => |ptr| ffi::gsk_render_node_ref(ptr),
+        unref => |ptr| ffi::gsk_render_node_unref(ptr),
+        get_type => || ffi::gsk_render_node_type_get_type(),
+    }
+}
+
+impl RenderNode {
+    //#[doc(alias = "gsk_render_node_deserialize")]
+    //pub fn deserialize(bytes: &glib::Bytes, error_func: /*Unimplemented*/FnMut(/*Ignored*/ParseLocation, /*Ignored*/ParseLocation, &glib::Error), user_data: /*Unimplemented*/Option<Fundamental: Pointer>) -> Option<RenderNode> {
+    //    unsafe { TODO: call ffi:gsk_render_node_deserialize() }
+    //}
+
+    pub fn downcast<T: IsARenderNode>(self) -> Result<T, RenderNode> {
+        unsafe {
+            if self.get_node_type() == T::NODE_TYPE {
+                Ok(from_glib_full(self.to_glib_full()))
+            } else {
+                Err(self)
+            }
+        }
+    }
+
+    pub fn downcast_ref<T: IsARenderNode>(&self) -> Option<&T> {
+        unsafe {
+            if self.get_node_type() == T::NODE_TYPE {
+                Some(&*(self as *const RenderNode as *const T))
+            } else {
+                None
+            }
+        }
+    }
+
+    pub fn draw(&self, cr: &cairo::Context) {
+        unsafe {
+            ffi::gsk_render_node_draw(self.to_glib_none().0, mut_override(cr.to_glib_none().0));
+        }
+    }
+
+    pub fn get_bounds(&self) -> graphene::Rect {
+        unsafe {
+            let mut bounds = graphene::Rect::uninitialized();
+            ffi::gsk_render_node_get_bounds(self.to_glib_none().0, bounds.to_glib_none_mut().0);
+            bounds
+        }
+    }
+
+    pub fn get_node_type(&self) -> RenderNodeType {
+        unsafe { from_glib(ffi::gsk_render_node_get_node_type(self.to_glib_none().0)) }
+    }
+
+    pub fn serialize(&self) -> Option<glib::Bytes> {
+        unsafe { from_glib_full(ffi::gsk_render_node_serialize(self.to_glib_none().0)) }
+    }
+
+    pub fn write_to_file(&self, filename: &str) -> Result<(), glib::Error> {
+        unsafe {
+            let mut error = ptr::null_mut();
+            let _ = ffi::gsk_render_node_write_to_file(
+                self.to_glib_none().0,
+                filename.to_glib_none().0,
+                &mut error,
+            );
+            if error.is_null() {
+                Ok(())
+            } else {
+                Err(from_glib_full(error))
+            }
+        }
+    }
+}
+
+pub trait IsARenderNode:
```

Comment:
Not a fan of the name. I would've called it `RenderNode` but that exists already. `RenderNodeType`? Not sure :)

Reply:
To be honest, I'm not a fan of the name either.  Not only does `RenderNode` exist, but also `RenderNodeType` also exists.  At first, I was using the name `RenderNodeSubType`.  Do you have any other suggestions?

Reply:
Maybe just `IsRenderNode` ?

Reply:
I mean, `IsRenderNode` sounds better than `IsARenderNode` to me :)